internal/stores/hostdb.go

```diff
@@ -181,3 +195,236 @@ func NewJSONHostDB(dir string) (*JSONHostDB, modules.ConsensusChangeID, error) {
+	}
+	for _, i := range h.Interactions {
+		hdbHost.Interactions = append(hdbHost.Interactions, i.Interaction())
+	}
+	return hdbHost
+}
+
+// Announcement converts a host into a hostdb.Announcement.
+func (a announcement) Announcement() hostdb.Announcement {
+	hostdbAnnouncement := hostdb.Announcement{
+		Index: consensus.ChainIndex{
+			Height: a.BlockHeight,
+		},
+		Timestamp:  a.Timestamp,
+		NetAddress: a.NetAddress,
+	}
+	copy(hostdbAnnouncement.Index.ID[:], a.BlockID)
+	return hostdbAnnouncement
+}
+
+// Interaction converts an interaction into a hostdb.Interaction.
+func (i interaction) Interaction() hostdb.Interaction {
+	return hostdb.Interaction{
+		Timestamp: i.Timestamp,
+		Type:      i.Type,
+		Result:    i.Result,
+	}
+}
+
+// NewSQLHostDB uses a given Dialector to connect to a SQL database.  NOTE: Only
+// pass migrate=true for the first instance of SQLHostDB if you connect via the
+// same Dialector multiple times.
+func NewSQLHostDB(conn gorm.Dialector, migrate bool) (*SQLHostDB, modules.ConsensusChangeID, error) {
+	db, err := gorm.Open(conn, &gorm.Config{})
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+
+	if migrate {
+		// Create the tables.
+		if err := db.AutoMigrate(&host{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&interaction{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&announcement{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+		if err := db.AutoMigrate(&consensusInfo{}); err != nil {
+			return nil, modules.ConsensusChangeID{}, err
+		}
+	}
+
+	// Get latest consensus change ID or init db.
+	var ci consensusInfo
+	err = db.Where(&consensusInfo{ID: consensusInfoID}).
+		Attrs(consensusInfo{
+			ID:   consensusInfoID,
+			CCID: modules.ConsensusChangeBeginning[:],
+		}).
+		FirstOrCreate(&ci).Error
+	if err != nil {
+		return nil, modules.ConsensusChangeID{}, err
+	}
+	var ccid modules.ConsensusChangeID
+	copy(ccid[:], ci.CCID)
+
+	return &SQLHostDB{
+		staticDB: db,
+	}, ccid, nil
+}
+
+// Host returns information about a host.
+func (db *SQLHostDB) Host(hostKey consensus.PublicKey) (hostdb.Host, error) {
+	var h host
+	tx := db.staticDB.Where(&host{PublicKey: hostKey[:]}).
+		Preload("Interactions").
+		Preload("Announcements").
+		Find(&h)
+	if tx.Error == nil && tx.RowsAffected == 0 {
+		return hostdb.Host{}, ErrHostNotFound
+	}
+	return h.Host(), tx.Error
+}
+
+// Hosts returns up to max hosts that have not been interacted with since
+// the specified time.
+func (db *SQLHostDB) Hosts(notSince time.Time, max int) ([]hostdb.Host, error) {
+	lastInteraction := func(h hostdb.Host) time.Time {
+		if len(h.Interactions) == 0 {
+			return time.Time{}
+		}
+		return h.Interactions[len(h.Interactions)-1].Timestamp
+	}
+	hosts, err := db.SelectHosts(math.MaxInt, func(h hostdb.Host) bool { return true })
```

Comment:
surely a relational DB allows us to do better than this, right? Can we add a `lastInteraction` column or something?

Reply:
Yep. Did some reading and updated the query. It's now a query that will fetch all the keys of hosts which fulfil the criteria from the db and then another one which fetches the full host information just for the returned keys.

Reply:
this host db is going to be so much fun writing queries for 😄